sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala

```diff
@@ -2079,4 +2078,621 @@ case class GpuConv(num: Expression, fromBase: Expression, toBase: Expression)
+          }
+        }
+        val decPartPad = withResource(decPartNegExp) { _ =>
+          decPartNegExp.pad(d, PadSide.RIGHT, "0")
+        }
+        val decPartNegSubstr = withResource(decPartPad) { _ =>
+          decPartPad.substring(0, d)
+        }
+        (getZeroCv(cv.getRowCount.toInt), decPartNegSubstr)
+    }
+  }
+
+  private def normalDoubleSplit(cv: ColumnVector, d: Int): (ColumnVector, ColumnVector) = {
+    val roundedStr = withResource(cv.round(d, RoundMode.HALF_EVEN)) { rounded =>
+      rounded.castTo(DType.STRING)
+    }
+    val (intPart, decPart) = withResource(roundedStr) { _ =>
+      withResource(roundedStr.stringSplit(".", 2)) { intAndDec =>
+        (intAndDec.getColumn(0).incRefCount(), intAndDec.getColumn(1).incRefCount())
+      }
+    }
+    val intPartNoNeg = withResource(intPart) { _ =>
+      removeNegSign(intPart)
+    }
+    val decPartPad = withResource(decPart) { _ =>
+      decPart.pad(d, PadSide.RIGHT, "0")
+    }
+    (intPartNoNeg, decPartPad)
+  }
+
+  private def expDoubleSplit(cv: ColumnVector, d: Int): (ColumnVector, ColumnVector) = {
+    // handle special case: 1.234567e+7 or 1.234567e-6
+    // get three parts first:
+    val replaceDelimToE = withResource(Scalar.fromString("e")) { e =>
+      withResource(Scalar.fromString(".")) { p =>
+        cv.stringReplace(e, p)
+      }
+    }
+    // get three parts: 1.234567e+7 -> 1, 234567, +7
+    val (intPartSign, decPart, expPart) = withResource(replaceDelimToE) { _ =>
+      withResource(replaceDelimToE.stringSplit(".", 3)) { intDecExp =>
+        (intDecExp.getColumn(0).incRefCount(),
+          intDecExp.getColumn(1).incRefCount(),
+          intDecExp.getColumn(2).incRefCount())
+      }
+    }
+    val intPart = withResource(intPartSign) { _ =>
+      removeNegSign(intPartSign)  
+    }
+    val exp = withResource(expPart) { _ =>
+      expPart.castTo(DType.INT32)
+    }
+    val (intPartPosExp, decPartPosExp) = closeOnExcept(intPart) { _ =>
+      closeOnExcept(decPart) { _ =>
+        closeOnExcept(exp) { _ =>
+          handleDoublePosExp(cv, intPart, decPart, exp, d)
+        }
+      }
+    }
+    val (intPartNegExp, decPartNegExp) = withResource(intPart) { _ =>
+      withResource(decPart) { _ =>
+        closeOnExcept(exp) { _ =>
+          handleDoubleNegExp(cv, intPart, decPart, exp, d)
+        }
+      }
+    }
+    val expPos = withResource(exp) { _ =>
+      withResource(Scalar.fromInt(0)) { zero =>
+        exp.greaterOrEqualTo(zero)
+      }
+    }
+    withResource(expPos) { _ =>
+      val intPartExp = withResource(intPartPosExp) { _ =>
+        withResource(intPartNegExp) { _ =>
+          expPos.ifElse(intPartPosExp, intPartNegExp)
+        }
+      }
+      val decPartExp = withResource(decPartPosExp) { _ =>
+        withResource(decPartNegExp) { _ =>
+          expPos.ifElse(decPartPosExp, decPartNegExp)
+        }
+      }
+      (intPartExp, decPartExp)
```

Comment:
If execption occurs at line 2371, then `intPartExp` will be leaked.
Should handle this resource pair as a whole.
```
      withResource(ArrayBuffer.empty[AutoCloseable]) { resource_array =>
        
        var res1 = genRes()
        resource_array ++= res1
        
        var res2 = genRes()
        resource_array ++= res2
        
      }
```